### ISSUE TITLE
Unrecoverable cluster after restart

### ISSUE THREAD

**Issue by quentez** [flynn/flynn#2760]
In order to perform system updates on the different nodes of a cluster, I restarted them one by one. I didn't wait for confirmation that the cluster was in a healthy state in-between restarts, assuming that it would recover eventually (I wasn't worried about a few minutes of downtime, since this wasn't a production system).

After all the nodes were restarted, the cluster never fully came back online. It was stuck trying to start the controller app on one of the nodes (after a while there was 10+ instances of the controller running on that node). I ended up having to boot a whole new cluster.

This is the main error that was logged again and again as the cluster was trying to recover: https://gist.github.com/quentez/2445a9fbbd854f16151637e8488961f5 It never moved past that state.

You can find the full logs for all the 4 nodes here: https://db.tt/m6BQnzWF

**Comment by lmars** [comment]
@quentez I've had a good look through the logs and I did find a couple of bugs (fixed in #2790) but these did not necessarily lead to the cluster not recovering.

The underlying issue was that, even though postgres jobs were running, the postgres cluster never recovered, so the controller was just hanging on startup [waiting for postgres](https://github.com/flynn/flynn/blob/v20160506.0/controller/controller.go#L61) (hence the continual `did not find any controller api instances` errors).

The initial reboot happened between `2016-04-21T04:13:55+0000` and `2016-04-21T04:15:35+0000`, and here are some postgres logs from each of the 4 hosts in the few minutes after that time:

### `Flynn0001`

```
 t=2016-04-21T04:15:59+0000 lvl=info msg="got discoverd state" app=postgres component=peer fn=handleDiscoverdState index=483 state=true
 t=2016-04-21T04:15:59+0000 lvl=info msg="starting state evaluation" app=postgres component=peer fn=evalClusterState
 t=2016-04-21T04:15:59+0000 lvl=info msg="assuming sync role" app=postgres component=peer role=sync fn=assumeSync
 t=2016-04-21T04:15:59+0000 lvl=info msg="starting state evaluation" app=postgres component=peer fn=evalClusterState
 t=2016-04-21T04:15:59+0000 lvl=info msg="reconfiguring database" app=postgres component=peer fn=applyConfig
 t=2016-04-21T04:15:59+0000 lvl=dbug msg="starting database" app=postgres component=peer fn=applyConfig
 t=2016-04-21T04:15:59+0000 lvl=info msg="starting up as standby" app=postgres component=postgres fn=assumeStandby upstream=100.100.65.2:5432
 t=2016-04-21T04:15:59+0000 lvl=info msg="waiting for upstream to come online" app=postgres component=postgres fn=waitForUpstream upstream=100.100.65.2:5432
 t=2016-04-21T04:15:59+0000 lvl=eror msg="heartbeat failed" component=discoverd service=postgres-api addr=100.100.59.3:80 err="unknown_error: Something went wrong"
 t=2016-04-21T04:15:59+0000 lvl=eror msg="heartbeat failed" component=discoverd service=postgres addr=100.100.59.11:5432 err="unknown_error: Something went wrong"
...
 t=2016-04-21T04:16:23+0000 lvl=eror msg="heartbeat failed" component=discoverd service=postgres addr=100.100.59.11:5432 err="unknown_error: Something went wrong"
 t=2016-04-21T04:16:23+0000 lvl=eror msg="heartbeat failed" component=discoverd service=postgres-api addr=100.100.59.3:80 err="unknown_error: Something went wrong"
 t=2016-04-21T04:16:29+0000 lvl=eror msg="error getting upstream status" app=postgres component=postgres fn=waitForUpstream upstream=100.100.65.2:5432 err="Get http://100.100.65.2:5433/status: dial tcp 100.100.65.2:5433: i/o timeout"
 t=2016-04-21T04:16:29+0000 lvl=eror msg="upstream did not come online in time" app=postgres component=postgres fn=waitForUpstream upstream=100.100.65.2:5432
 t=2016-04-21T04:16:29+0000 lvl=eror msg="error applying database config" app=postgres component=peer fn=applyConfig err="upstream is offline"
 t=2016-04-21T04:16:29+0000 lvl=info msg="discoverd peers" app=postgres component=peer peers=1
 t=2016-04-21T04:16:29+0000 lvl=info msg="reconfiguring database" app=postgres component=peer fn=applyConfig
 t=2016-04-21T04:16:29+0000 lvl=dbug msg="starting database" app=postgres component=peer fn=applyConfig
 t=2016-04-21T04:16:29+0000 lvl=info msg="starting up as standby" app=postgres component=postgres fn=assumeStandby upstream=100.100.65.2:5432
 t=2016-04-21T04:16:29+0000 lvl=info msg="waiting for upstream to come online" app=postgres component=postgres fn=waitForUpstream upstream=100.100.65.2:5432
 t=2016-04-21T04:16:32+0000 lvl=eror msg="error getting upstream status" app=postgres component=postgres fn=waitForUpstream upstream=100.100.65.2:5432 err="Get http://100.100.65.2:5433/status: dial tcp 100.100.65.2:5433: no route to hos
 t=2016-04-21T04:16:35+0000 lvl=eror msg="error getting upstream status" app=postgres component=postgres fn=waitForUpstream upstream=100.100.65.2:5432 err="Get http://100.100.65.2:5433/status: dial tcp 100.100.65.2:5433: no route to hos
 t=2016-04-21T04:16:38+0000 lvl=eror msg="error getting upstream status" app=postgres component=postgres fn=waitForUpstream upstream=100.100.65.2:5432 err="Get http://100.100.65.2:5433/status: dial tcp 100.100.65.2:5433: no route to hos
 t=2016-04-21T04:16:41+0000 lvl=eror msg="error getting upstream status" app=postgres component=postgres fn=waitForUpstream upstream=100.100.65.2:5432 err="Get http://100.100.65.2:5433/status: dial tcp 100.100.65.2:5433: no route to hos
 t=2016-04-21T04:16:41+0000 lvl=eror msg="upstream did not come online in time" app=postgres component=postgres fn=waitForUpstream upstream=100.100.65.2:5432
...
 t=2016-04-21T04:16:41+0000 lvl=info msg="starting state evaluation" app=postgres component=peer fn=evalClusterState
 t=2016-04-21T04:16:41+0000 lvl=warn msg="would takeover but no async peers present" app=postgres component=peer fn=startTakeover reason="primary gone" min_wal=0/19A59870
 t=2016-04-21T04:16:41+0000 lvl=info msg="reconfiguring database" app=postgres component=peer fn=applyConfig
 t=2016-04-21T04:16:41+0000 lvl=dbug msg="starting database" app=postgres component=peer fn=applyConfig
 t=2016-04-21T04:16:41+0000 lvl=info msg="starting up as standby" app=postgres component=postgres fn=assumeStandby upstream=100.100.65.2:5432
 t=2016-04-21T04:16:41+0000 lvl=info msg="waiting for upstream to come online" app=postgres component=postgres fn=waitForUpstream upstream=100.100.65.2:5432
 t=2016-04-21T04:16:53+0000 lvl=eror msg="error getting upstream status" app=postgres component=postgres fn=waitForUpstream upstream=100.100.65.2:5432 err="Get http://100.100.65.2:5433/status: dial tcp 100.100.65.2:5433: no route to hos
 t=2016-04-21T04:16:53+0000 lvl=eror msg="upstream did not come online in time" app=postgres component=postgres fn=waitForUpstream upstream=100.100.65.2:5432
...
 t=2016-04-21T04:20:20+0000 lvl=dbug msg="starting database" app=postgres component=peer fn=applyConfig
 t=2016-04-21T04:20:20+0000 lvl=info msg="starting up as standby" app=postgres component=postgres fn=assumeStandby upstream=100.100.65.2:5432
 t=2016-04-21T04:20:20+0000 lvl=info msg="waiting for upstream to come online" app=postgres component=postgres fn=waitForUpstream upstream=100.100.65.2:5432
 panic: dial tcp 100.100.65.2:5432: connection refused [recovered
        panic: dial tcp 100.100.65.2:5432: connection refused
```

### `Flynn0002`

No postgres jobs running in the minutes after the reboot.

### `Flynn0003`

No postgres jobs running in the minutes after the reboot.

### `Flynn0004`

```
 t=2016-04-21T04:17:36+0000 lvl=dbug msg=moving app=postgres component=peer
 t=2016-04-21T04:17:36+0000 lvl=info msg="starting state evaluation" app=postgres component=peer fn=evalClusterState
 t=2016-04-21T04:17:36+0000 lvl=info msg="assuming unassigned role" app=postgres component=peer role=unassigned fn=assumeUnassigned
 t=2016-04-21T04:17:36+0000 lvl=info msg="reconfiguring database" app=postgres component=peer fn=applyConfig
 t=2016-04-21T04:17:36+0000 lvl=dbug msg="skipping stop, already offline" app=postgres component=peer fn=applyConfig
...
 t=2016-04-21T04:17:36+0000 lvl=info msg="discoverd peers" app=postgres component=peer peers=2
 t=2016-04-21T04:17:36+0000 lvl=dbug msg=moving app=postgres component=peer
 t=2016-04-21T04:17:36+0000 lvl=info msg="starting state evaluation" app=postgres component=peer fn=evalClusterState
 t=2016-04-21T04:17:36+0000 lvl=dbug msg="at rest" app=postgres component=peer fn=rest
 t=2016-04-21T04:17:36+0000 lvl=eror msg="error getting postgres info" app=postgres component=http err="postgres is not running"
 t=2016-04-21T04:17:36+0000 lvl=eror msg="error getting postgres info" app=postgres component=http err="postgres is not running"
 t=2016-04-21T04:17:36+0000 lvl=eror msg="error getting postgres info" app=postgres component=http err="postgres is not running"
...
<the above repeated until the process exits>
...
 t=2016-04-21T04:24:36+0000 lvl=eror msg="error getting postgres info" app=postgres component=http err="postgres is not running"
 t=2016-04-21T04:24:36+0000 lvl=eror msg="error getting postgres info" app=postgres component=http err="postgres is not running"
 t=2016-04-21T04:24:36+0000 lvl=eror msg="error getting postgres info" app=postgres component=http err="postgres is not running"
 t=2016-04-21T04:24:36+0000 lvl=info msg="command exited" app=containerinit fn=containerInitApp status=0

```

I was initially skeptical as to why there were only 2 postgres jobs, but that is [how many the host fixer starts](https://github.com/flynn/flynn/blob/v20160506.0/host/fixer/fixer.go#L467-L468) and should be sufficient to get a read-write cluster. 

So we have a few questions we need to answer:

* why did the first postgres job remain a `sync` and continue to wait for a non-existent upstream (`100.100.65.2`) to come online?
* why did the second postgres job assume an unassigned role and not start postgres?
* why did the host fixer not realise postgres was not read-write and rectify the situation?

/cc @titanous @josephglanville

**Comment by titanous** [comment]
Flynn is unmaintained and our infrastructure will shut down on June 1, 2021. See [the README](https://github.com/flynn/flynn#flynn-is-unmaintained) for details.
